tests/scripts/thread-cert/test_netdata_publisher.py

```diff
@@ -0,0 +1,270 @@
+        self.assertEqual(service[3], 's')
+
+    def verify_anycast_services(self, services):
+        # Verify a list of anycast `services` from `get_services()`
+        for service in services:
+            self.verify_anycast_service(service)
+
+    def verify_unicast_service(self, service):
+        # Verify the data in a single unicast `service` from `get_services()`
+        # Example of `service`: ['44970', '5d', 'fd000db800000000c6b0e5ee81f940e8223d', 's', '7000']
+        self.assertEqual(int(service[0]), THREAD_ENTERPRISE_NUMBER)
+        # Check service data
+        service_data = parse_hex_string(service[1])
+        self.assertTrue(len(service_data) >= 1)
+        self.assertEqual(service_data[0], UNICAST_SERVICE_NUM)
+        # Verify that it stable
+        self.assertEqual(service[3], 's')
+
+    def verify_unicast_services(self, services):
+        # Verify a list of unicast `services` from `get_services()`
+        for service in services:
+            self.verify_unicast_service(service)
+
+    def test(self):
+        leader = self.nodes[LEADER]
+        router1 = self.nodes[ROUTER1]
+        router2 = self.nodes[ROUTER2]
+        router3 = self.nodes[ROUTER3]
+        router4 = self.nodes[ROUTER4]
+        router5 = self.nodes[ROUTER5]
+        end_dev1 = self.nodes[END_DEV1]
+        end_dev2 = self.nodes[END_DEV2]
+        end_dev3 = self.nodes[END_DEV3]
+        end_dev4 = self.nodes[END_DEV4]
+        end_dev5 = self.nodes[END_DEV5]
+
+        nodes = self.nodes.values()
+        routers = [router1, router2, router3, router4, router5]
+        end_devs = [end_dev1, end_dev2, end_dev3, end_dev4, end_dev5]
+
+        # Start the nodes
+
+        leader.start()
+        self.simulator.go(5)
+        self.assertEqual(leader.get_state(), 'leader')
+
+        for router in routers:
+            router.start()
+            self.simulator.go(5)
+            self.assertEqual(router.get_state(), 'router')
+
+        for end_dev in end_devs:
+            end_dev.start()
+            self.simulator.go(5)
+            self.assertEqual(end_dev.get_state(), 'child')
+
+        #---------------------------------------------------------------------------------
+        # DNS/RSP anycast entries
+
+        # Publish DNS/SRP anycast on leader and all routers (6 nodes).
+
+        leader.netdata_publish_dnssrp_anycast(ANYCAST_SEQ_NUM)
+        for node in routers:
+            node.netdata_publish_dnssrp_anycast(ANYCAST_SEQ_NUM)
+        self.simulator.go(WAIT_TIME)
+
+        # Check all entries are present in the network data
+
+        services = leader.get_services()
+        self.assertEqual(len(services), min(1 + len(routers), DESIRED_NUM_ANYCAST))
+        self.verify_anycast_services(services)
+
+        # Publish same entry on all end-devices (5 nodes).
+
+        for node in end_devs:
+            node.netdata_publish_dnssrp_anycast(ANYCAST_SEQ_NUM)
+            print(node.name)
+        self.simulator.go(WAIT_TIME)
```

Comment:
may move publish on all end-devices before on all leader and routers, and add verification that leader&routers are preferred over end devices when exceeding the limitation?

Reply:
The preferred entry logic comes into picture when we exceed the desired number and want to remove some entries. There is no easy and reliable way in the test to create a situation where the number of entries exceed the limit. It is due to  how nice the algo works :)... each node picks random wait time and then as soon they see we have enough entries, they stop adding their own entry.

I had this situation covered in the test when I had the "on-mesh/external route" prefix support included. But it is removed. There I could use the "preference" (high/med/low) of prefix entries to create a situation where total number entries would exceed the limit. We cannot use the same with service entries (there is no "preference" in service entries).

Reply:
There may be a way around this by manually adding similar "service" entries directly on some node. It's a bit hacky though.

Reply:
Added a new test case (in a new commit) where check the when removing entries, publisher prefers and keeps the entries from routers over ones from end-devices.

Reply:
> There may be a way around this by manually adding similar "service" entries directly on some node. It's a bit hacky though.
The workaround in this PR does sound good to me. the other possible option might be add API to allow configure different publisher entries on child/router, just like what Thread 1.1 Test Harness do for TC 5.2.3, which test maximal active router capacity, by configuring routerupgradethreshold/routerdowngradethreshold.
Such API and CLI command might be required if adding similar active SRP Server capacity test.

Reply:
That can also work, but in genera, I would prefer and suggest to avoid adding new APIs for testing purposes only if/when we can. The main reason is I think it would be better that the firmware we use for testing to be closest to what we intend to use/ship in products.